src/playercards/spawnbag/SpawnBag.ttslua

```diff
@@ -0,0 +1,202 @@
+  recallZone = saveTable.recall
+end
+
+-- Generates a table of save state that can be included in the parent object's onSave
+SpawnBag.getStateForSave = function()
+  local saveState = {
+    placed = placedSpecs,
+    placedObjects = placedObjectGuids,
+    recall = recallZone,
+  }
+
+  return saveState
+end
+
+-- Places the given spawnSpec on the table.  See SpawnBag.ttslua header for spawnSpec table data and
+-- examples
+SpawnBag.spawn = function(spawnSpec)
+  -- Limit to one placement at a time
+  if (placedSpecs[spawnSpec.name]) then
+    return
+  end
+  if (spawnSpec == nil) then
+    -- TODO: error here
+    return
+  end
+  local cardsToSpawn = { }
+  local allCardsBag = getObjectFromGUID(ALL_CARDS_GUID)
+  for _, cardId in ipairs(spawnSpec.cards) do
+    local cardData = allCardsBag.call("getCardById", { id = cardId })
+    if (cardData ~= nil) then
+      table.insert(cardsToSpawn, cardData)
+    else
+      -- TODO: error here
+    end
+  end
+  if (spawnSpec.spread) then
+    Spawner.spawnCardSpread(cardsToSpawn, spawnSpec.globalPos, spawnSpec.rotation, false, recordPlacedObject)
+  else
+    Spawner.spawnCards(cardsToSpawn, spawnSpec.globalPos, spawnSpec.rotation, false, recordPlacedObject)
+  end
+  placedSpecs[spawnSpec.name] = true
+end
+
+-- Recalls all spawned objects to the bag, and clears the placedObjectGuids list
+SpawnBag.recall = function()
+  local trash = spawnObjectData({data = RECALL_BAG, position = self.getPosition()})
+  for guid, _ in pairs(placedObjectGuids) do
+    local obj = getObjectFromGUID(guid)
+    if (obj ~= nil) then
+      if (isInRecallZone(obj)) then
+        trash.putObject(obj)
+      end
+      placedObjectGuids[guid] = nil
+    end
+  end
+
+  trash.destruct()
+  -- We've recalled everything we can, some cards may have been moved out of the
+  -- card area.  Just reset at this point.
+  placedSpecs = { }
+  placedObjectGuids = { }
+  recallZone = nil
+end
+
+-- Callback for when an object has been spawned.  Tracks the object for later recall and updates the
+-- recall zone.
+function recordPlacedObject(spawned)
+  placedObjectGuids[spawned.getGUID()] = true
+  expandRecallZone(spawned)
+end
+
+-- Expands the current recall zone based on the position of the given object.  The recall zone will
+-- be maintained as the bounding box of the extreme object positions, plus a small amount of buffer
+function expandRecallZone(spawnedCard)
+  local pos = spawnedCard.getPosition()
+  if (recallZone == nil) then
+    -- First card out of the bag, initialize surrounding that
+    recallZone = { }
+    recallZone.upperLeft = { x = pos.x + RECALL_BUFFER_X, z = pos.z + RECALL_BUFFER_Z }
+    recallZone.lowerRight = { x = pos.x - RECALL_BUFFER_X, z = pos.z - RECALL_BUFFER_Z }
+    return
+  else
+    if (pos.x > recallZone.upperLeft.x) then
+      recallZone.upperLeft.x = pos.x + RECALL_BUFFER_X
+    end
+    if (pos.x < recallZone.lowerRight.x) then
+      recallZone.upperLeft.x = pos.x - RECALL_BUFFER_X
```

Comment:
pretty sure this should contain ".lowerRight" instead of "upperLeft"

Reply:
It should, thanks for the catch.  Hazard of test case being only one line of cards.